src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,139 @@
+import React, { useRef, useState } from 'react';
+import { Todo } from '../../types/Todo';
+import cn from 'classnames';
+import { USER_ID } from '../../api/todos';
+import { useTodoActions } from '../../TodoContext';
+// import { Errors } from '../../constants/Errors';
+// import { useErrorNotificationApi } from '../Error/ErrorContext';
+
+type Props = {
+  todo: Todo;
+  isProcessed: boolean;
+  isSubmitting: boolean;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo, isProcessed }) => {
+  const { title, id, completed } = todo;
+  //eslint-disable-next-line
+  const { handleDeleteTodo, handleUpdateTodo, handleToggleComplete } =
+    useTodoActions();
+  const [newTitle, setNewTitle] = useState('');
+  const [isEditinigTitle, setIsEditinigTitle] = useState<boolean>(false);
+
+  const inputElement = useRef<HTMLInputElement | null>(null);
+
+  function handleMarkComplete() {
+    handleToggleComplete(todo);
+  }
+
+  const handleEditTitleSave = async (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (!isEditinigTitle) {
+      return;
+    }
+
+    let isSuccessful = true;
+
+    const newTrimmedTitle = newTitle.trim();
+
+    if (!newTrimmedTitle.length) {
+      isSuccessful = await handleDeleteTodo(id);
+    } else if (newTrimmedTitle !== title) {
+      isSuccessful = await handleUpdateTodo({
+        id: id,
+        title: newTrimmedTitle,
+        userId: USER_ID,
+        completed: completed,
+      });
+    }
+
+    if (isSuccessful) {
+      setNewTitle('');
+      setIsEditinigTitle(false);
+    } else {
+      inputElement.current?.focus();
+    }
+  };
+
+  const handleEditTitleStart = () => {
+    setIsEditinigTitle(true);
+    setNewTitle(title);
+    if (inputElement.current) {
+      inputElement.current.focus();
+    }
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    if (event.key === 'Escape') {
+      setIsEditinigTitle(false);
+      setNewTitle(title);
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={cn('todo', {
+        completed: completed,
+      })}
+    >
+      {/* eslint-disable jsx-a11y/label-has-associated-control */}
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={completed}
+          onChange={handleMarkComplete}
+        />
+      </label>
+
+      {isEditinigTitle ? (
+        <form onSubmit={handleEditTitleSave}>
+          <input
+            data-cy="TodoTitleField"
+            type="text"
+            className="todo__title-field"
+            placeholder="Empty todo will be deleted"
+            value={newTitle}
+            onChange={e => setNewTitle(e.currentTarget.value)}
```

Comment:
handler function

Reply:
you forget to move function to a separate variable 🙂